File: src/main/java/ru/johnspade/service/PostService.java
package ru.johnspade.service;

import com.google.common.base.Optional;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import ru.johnspade.dao.Post;
import ru.johnspade.dao.Tag;
import ru.johnspade.repository.PostRepository;
import ru.johnspade.web.NodeState;
import ru.johnspade.web.Rss;
import ru.johnspade.web.Tree;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.List;
import java.util.Locale;

@Service
public class PostService {

	private final static int PAGE_SIZE = 4;
	private final static String SORT_PROPERTY = "date";
	private final static Locale LOCALE = Locale.forLanguageTag("ru");
	@Autowired
	private CacheService cacheService;
	@Autowired
	PostRepository repository;
	@Autowired
	private Rss rss;

	@Cacheable(cacheNames = "postExists", keyGenerator = "cacheKeyGenerator")
	public boolean exists(int id) {
		return repository.exists(id);
	}

	@Cacheable(cacheNames = "post", keyGenerator = "cacheKeyGenerator")
	public Post get(int id) {
		return repository.getOne(id);
	}

	public Post save(Post post) {
		cacheService.setlastSaveTimestamp(new Date());
		return repository.save(post);
	}

	public void delete(Post post) {
		cacheService.setlastSaveTimestamp(new Date());
		repository.delete(post);
	}

	@Cacheable(cacheNames = "mostRecentPost", keyGenerator = "cacheKeyGenerator")
	public Optional<Post> findMostRecent() {
		Post post = null;
		Page<Post> result = repository.findAll(new PageRequest(0, 1, Sort.Direction.DESC, SORT_PROPERTY));
		if (result.getNumberOfElements() > 0)
			post = result.getContent().get(0);
		return Optional.fromNullable(post);
	}

	@Cacheable(cacheNames = "posts", keyGenerator = "cacheKeyGenerator")
	public List<Post> findAll() {
		List<Post> posts = repository.findAll();
		Collections.sort(posts, new Comparator<Post>() {
			@Override
			public int compare(Post o1, Post o2) {
				return o2.getDate().compareTo(o1.getDate());
			}
		});
		return Collections.unmodifiableList(posts);
	}

	@Cacheable(cacheNames = "postsPage", keyGenerator = "cacheKeyGenerator")
	public Page<Post> findAll(int pageNumber) {
		return repository.findAll(new PageRequest(pageNumber, PAGE_SIZE, Sort.Direction.DESC, SORT_PROPERTY));
	}

	public Page<Post> findAllByTag(Tag tag, int pageNumber) {
		return repository.findAllByTag(tag, new PageRequest(pageNumber, PAGE_SIZE, Sort.Direction.DESC, SORT_PROPERTY));
	}

	public Page<Post> findRecent() {
		return repository.findAll(new PageRequest(0, 25, Sort.Direction.DESC, SORT_PROPERTY));
	}

	@Cacheable(cacheNames = "rss", keyGenerator = "cacheKeyGenerator")
	public Rss getRss() {
		return rss;
	}

	@Cacheable(cacheNames = "tree", keyGenerator = "cacheKeyGenerator")
	public List<Tree> getTree() {
		List<Tree> years = new ArrayList<>();
		List<Post> posts = findAll();
		for (Post post : posts) {
			Calendar cal = Calendar.getInstance();
			cal.setTime(post.getDate());
			boolean yearFound = false;
			for (Tree year : years) {
				if (Integer.toString(cal.get(Calendar.YEAR)).equals(year.getText())) {
					boolean monthFound = false;
					List<Tree> months = year.getNodes();
					for (Tree month : months) {
						if (getMonthName(cal).equals(month.getText())) {
							addPostToMonth(post, month);
							monthFound = true;
							break;
						}
					}
					if (!monthFound) {
						Tree month = new Tree(getMonthName(cal));
						months.add(month);
						addPostToMonth(post, month);
					}
					yearFound = true;
					break;
				}
			}
			if (!yearFound) {
				Tree year = new Tree(Integer.toString(cal.get(Calendar.YEAR)));
				years.add(year);
				Tree month = new Tree(getMonthName(cal));
				years.get(years.size() - 1).getNodes().add(month);
				addPostToMonth(post, month);
			}
		}
		return years;
	}

	private String getMonthName(Calendar cal) {
		return cal.getDisplayName(Calendar.MONTH, Calendar.LONG, LOCALE);
	}

	private void addPostToMonth(Post post, Tree month) {
		month.getNodes().add(new Tree(post.getTitle(), new NodeState(), null, "/posts/" + Integer.toString(post.getId())));
		if (month.getTags() == null)
			month.setTags(Collections.singletonList("0"));
		month.setTags(Collections.singletonList(Integer.toString(Integer.parseInt(month.getTags().get(0)) + 1)));
	}

}

File: src/main/java/ru/johnspade/service/PostService.kt
package ru.johnspade.service

import com.google.common.base.Optional
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.cache.annotation.Cacheable
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.stereotype.Repository
import org.springframework.stereotype.Service
import ru.johnspade.dao.Post
import ru.johnspade.dao.Tag
import ru.johnspade.web.NodeState
import ru.johnspade.web.Tree
import java.util.*

@Repository
interface PostRepository : JpaRepository<Post, Int> {

	@Query("select p from Post p join p.tags t where t = ?1")
	fun findAllByTag(tag: Tag, pageable: Pageable): Page<Post>

}

@Service
open class PostService @Autowired constructor(
		private val cacheService: CacheService,
		private val repository: PostRepository
) {

	@Cacheable(cacheNames = arrayOf("postExists"), keyGenerator = "cacheKeyGenerator")
	open fun exists(id: Int): Boolean {
		return repository.exists(id)
	}

	@Cacheable(cacheNames = arrayOf("post"), keyGenerator = "cacheKeyGenerator")
	open operator fun get(id: Int): Post {
		return repository.getOne(id)
	}

	open fun save(post: Post): Post {
		cacheService.setlastSaveTimestamp(Date())
		return repository.save(post)
	}

	open fun delete(post: Post) {
		cacheService.setlastSaveTimestamp(Date())
		repository.delete(post)
	}

	@Cacheable(cacheNames = arrayOf("mostRecentPost"), keyGenerator = "cacheKeyGenerator")
	open fun findMostRecent(): Optional<Post> {
		var post: Post? = null
		val result = repository.findAll(PageRequest(0, 1, Sort.Direction.DESC, SORT_PROPERTY))
		if (result.numberOfElements > 0)
			post = result.content[0]
		return Optional.fromNullable(post)
	}

	@Cacheable(cacheNames = arrayOf("posts"), keyGenerator = "cacheKeyGenerator")
	open fun findAll(): List<Post> {
		val posts = repository.findAll()
		Collections.sort(posts) { o1, o2 -> o2.date.compareTo(o1.date) }
		return Collections.unmodifiableList(posts)
	}

	@Cacheable(cacheNames = arrayOf("postsPage"), keyGenerator = "cacheKeyGenerator")
	open fun findAll(pageNumber: Int): Page<Post> {
		return repository.findAll(PageRequest(pageNumber, PAGE_SIZE, Sort.Direction.DESC, SORT_PROPERTY))
	}

	open fun findAllByTag(tag: Tag, pageNumber: Int): Page<Post> {
		return repository.findAllByTag(tag, PageRequest(pageNumber, PAGE_SIZE, Sort.Direction.DESC, SORT_PROPERTY))
	}

	open fun findRecent(): Page<Post> {
		return repository.findAll(PageRequest(0, 25, Sort.Direction.DESC, SORT_PROPERTY))
	}

	@Cacheable(cacheNames = arrayOf("tree"), keyGenerator = "cacheKeyGenerator")
	open fun getTree(postDate: Calendar?): List<Tree> {
		val years = ArrayList<Tree>()
		val posts = findAll()
		posts.forEach { post ->
			val cal = Calendar.getInstance()
			cal.time = post.date
			val expandYear = postDate != null && postDate.get(Calendar.YEAR) == cal.get(Calendar.YEAR)
			val expandMonth = expandYear && postDate!!.get(Calendar.MONTH) == cal.get(Calendar.MONTH)
			var year = years.firstOrNull { it.text == cal.get(Calendar.YEAR).toString() }
			if (year == null) {
				year = Tree(cal.get(Calendar.YEAR).toString(), expandYear)
				years.add(year)
			}
			var month = year.nodes.firstOrNull { it.text == getMonthName(cal) }
			if (month == null) {
				month = Tree(getMonthName(cal), expandMonth)
				year.nodes.add(month)
			}
			addPostToMonth(post, month, expandMonth)
		}
		return years
	}

	private fun getMonthName(cal: Calendar): String {
		return cal.getDisplayName(Calendar.MONTH, Calendar.LONG, LOCALE)
	}

	private fun addPostToMonth(post: Post, month: Tree, expanded: Boolean) {
		month.expand(expanded)
		month.nodes.add(Tree(post.title, NodeState(false), null, "/posts/" + Integer.toString(post.id)))
		if (month.tags == null)
			month.tags = listOf("0")
		month.tags = listOf(Integer.toString(Integer.parseInt(month.tags[0]) + 1))
	}

	companion object {
		private val PAGE_SIZE = 4
		private val SORT_PROPERTY = "date"
		private val LOCALE = Locale.forLanguageTag("ru")
	}

}
